File: 202.快乐数.py
#
# @lc app=leetcode.cn id=202 lang=python3
#
# [202] 快乐数
#

# @lc code=start


class Solution:
    def isHappy(self, n: int) -> bool:
        def h(num):
            if num == 1:
                return True
            if num in s:
                return False
            s.add(num)
            new_num = 0
            li = list(str(num))
            for i in li:
                new_num += int(i)**2
            return h(new_num)
        s = set()
        return h(n)

# @lc code=end

File: 202.快乐数.java
/*
 * @lc app=leetcode.cn id=202 lang=java
 *
 * [202] 快乐数
 */

// @lc code=start
import java.util.HashSet;
class Solution {
    private HashSet<Integer> set = new HashSet<>();
    public boolean isHappy(int n) {
        if (n == 1) 
            return true;
        if (set.contains(n))
            return false;
        set.add(n);
        int nNew = 0;
        String nStr = n+"";
        for (int i = 0; i < nStr.length(); i++) {
            nNew += Math.pow(nStr.charAt(i)-'0', 2);
        }
        return isHappy(nNew);
    }
}
// @lc code=end

